man/citgm.1
.\" Manpage for citgm
.\" Contact jasnell@gmail.com to correct errors or typos
.TH "0.1.0" "MIT"
.SH NAME
citgm \- Canary in the Goldmine
.SH SYNOPSIS
citgm [options] <module>
.SH DESCRIPTION
citgm is a tool for testing Node.js modules against different Node.js builds.
Specifically, it automates the 'npm install' and 'npm test' steps and reports
back the status.
.SH OPTIONS
.TP
.BR \-h ", " \-\-help
output usage information
.TP
.BR \-V ", " \-\-version
output the version number
.TP
.BR \-v ", " \-\-verbose ", " \-\-loglevel " " \fI[level]\fR
Verbose output (silly,verbose,info,warn,error)
.TP
.BR \-q ", " \-\-npm-loglevel " " \fI[level]\fR
Verbose output (silent, error, warn, http, info, verbose, silly)
.TP
.BR \-l ", " \-\-lookup " " \fI<path>\fR
Use the lookup table provided at <path>
.TP
.BR \-d ", " \-\-nodedir " " \fI<path>\fR
Path to the node source to use when compiling native addons
.TP
.BR \-p ", " \-\-test-path " " \fI<path>\fR
Path to prepend to $PATH when running tests
.TP
.BR \-n ", " \-\-no-color
Turns off colorized output
.TP
.BR \-s ", " \-\-su
Allow running the tool as root
.TP
.BR \-m ", " \-\-markdown
Output results in markdown
.TP
.BR \-t ", " \-\-tap " " \fI[path]\fR
Output results in tap
.TP
.BR \-x ", " \-\-junit " " \fI[path]\fR
Output results in junit xml with optional file path
.TP
.BR \-o ", " \-\-timeout " " \fI<length>\fR
Set timeout for npm install
.TP
.BR \-c ", " \-\-sha " " \fI<commit-sha>\fR
Install module from commit-sha
.TP
.BR \-u ", " \-\-uid " " \fIuid\fR
Set the uid (posix only)
.TP
.BR \-g ", " \-\-gid " " \fIuid\fR
Set the gid (posix only)
.TP
.BR \-a ", " \-\-append
Turns on append results to file mode rather than replace
.TP
.BR \-\-tmpDir " " \fI<path>\fR
Directory to test modules in
.SH EXAMPLES
Test the latest underscore module or a specific version:

  citgm underscore@latest
  citgm underscore@1.3.0

Test a local module:

  citgm ./my-module

Test using a tar.gz from Github:

  citgm http://github.com/jasnell/activitystrea.ms/archive/master.tar.gz

.SH SEE ALSO
citgm-all
.SH AUTHOR
James M Snell <jasnell@gmail.com>
.SH COPYRIGHT
Copyright (c) 2015 citgm contributors
.SH LICENSE
The MIT License
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


RELEASES.md
# CITGM Release Process

This document describes the technical aspects of the citgm release process. The intended audience is those who have been authorized to create official releases for citgm, hosted on npm.

## Who can make a release?

Any member of the @nodejs/citgm team can make a release.

## How to create a release

1. Ensure your branch is up to date with upstream

```bash
$ git checkout master
$ git remote update -p
$ git reset --hard upstream/master
$ git diff upstream/master # this should be a no-op
```

2. Bump the version and create tag with semver-sync

```bash
$ npm version [<newversion> | major | minor | patch]
```

3. Push to github

```bash
$ git push upstream master --follow-tags
```

4. Publish to npm

```bash
$ npm publish
```


lib/citgm.js
'use strict';
const EventEmitter = require('events').EventEmitter;
const util = require('util');

const async = require('async');
const npa = require('npm-package-arg');
let which = require('which'); // Mocked in tests

const grabModuleData = require('./grab-module-data');
const grabProject = require('./grab-project');
const lookup = require('./lookup');
const npm = require('./npm');
const tempDirectory = require('./temp-directory');
const unpack = require('./unpack');

const windows = (process.platform === 'win32');
exports.windows = windows;

/**
 * The process here is straightforward:
 * 1. We pull down information about the module from npm
 * 2. We create a temporary working directory and pull down the
 *    module using the git repo shown in the npm metadata or,
 *    if the repo is not specified, by pulling the code from
 *    npm itself
 * 3. If npm-test is implemented, invoke to run the tests. Otherwise,
 *    if options.script is provided, run it as a child process.
 * 4. Filter the output of 4 through the Reporter script (details TBD)
 * 5. Output the results.
 **/

function find(app, context, next) {
  which(app, function(err, resolved) {
    if (err) {
      next(Error(app + ' not found in path!'));
      return;
    }
    context.emit('data', 'verbose', context.module.name + ' using-' + app,
        resolved);
    next(null, context);
  });
}

function init(context, next) {
  // Script is passed as an option to citgm, but it needs to also apply to the
  // Single module that citgm is testing.
  if (context.options.script) {
    context.module.script = context.options.script;
  }
  if (!windows) {
    if (context.options.uid)
      context.emit('data', 'verbose', context.module.name + ' using-uid',
          context.options.uid);
    if (context.options.gid)
      context.emit('data', 'verbose', context.module.name + ' using-gid',
          context.options.gid);
  }
  context.emit('data', 'silly', context.module.name + ' init-detail',
      context.module);
  next(null, context); // Inject the context
}

function sha1(mod) {
  const crypto = require('crypto');
  const shasum = crypto.createHash('sha1');
  shasum.update(mod);
  return shasum.digest('hex');
}

function extractDetail(mod) {
  const detail = npa(mod); // Will throw on failure
  detail.name = detail.name || sha1(mod);
  return detail;
}

/**
 * Setup the test
 *  - module = the name of the npm module to check out
 *  - options.script = path to a test script to run
 *  - options.reporter = path to a reporting script to run
 *
 * The Tester will emit specific events once it is run:
 *  - end = emitted when the test is complete
 *  - fail = emitted if the test fails for any reason
 *  - info = informational events (for debugging)
 *  - start = emitted when run is started
 **/

function Tester(mod, options) {
  if (!(this instanceof Tester))
    return new Tester(mod, options);
  EventEmitter.call(this);
  this.module = extractDetail(mod);
  this.options = options;
  this.testOutput = '';
  this.testError = '';
  this.cleanexit = false;
}

util.inherits(Tester, EventEmitter);

Tester.prototype.run = function() {
  this.emit('start', this.module.raw, this.options);

  async.waterfall([
    init.bind(null, this),
    find.bind(null, 'node'),
    find.bind(null, 'npm'),
    tempDirectory.create,
    grabModuleData,
    lookup,
    grabProject,
    unpack,
    npm.install,
    npm.test
  ], (err) => {
    if (!this.cleanexit) {
      const payload = {
        name: this.module.name || this.module.raw,
        version: this.module.version,
        flaky: this.module.flaky,
        expectFail: this.module.expectFail
      };
      if (err && !payload.expectFail) {
        this.emit('fail', err);
        payload.error = err;
      } else if (!err && payload.expectFail) {
        this.emit('fail', 'this module should have failed');
        payload.error = 'this module should have failed';
      }
      if (this.testOutput !== '') {
        payload.testOutput += this.testOutput + '\n';
      }
      if (this.testError !== '') {
        payload.testOutput += this.testError + '\n';
      }
      tempDirectory.remove(this, () => {
        this.emit('end', payload);
        this.cleanexit = true;
      });
    }
  });
};

Tester.prototype.cleanup = function () {
  this.cleanexit = true;
  const payload = {
    name: this.module.name || this.module.raw,
    error: Error('Process Interrupted')
  };
  this.emit('fail', payload.error);
  tempDirectory.remove(this, () => {
    this.emit('end', payload);
  });
};

exports.Tester = Tester;
